write.go

```diff
@@ -221,28 +233,47 @@ func writeProduceRequestV2(w *bufio.Writer, correlationID int32, clientID, topic
 	writeInt32(w, partition)
 	writeInt32(w, size)
 
-	const magicByte = 1
-	const attributes = 0
-
 	for _, msg := range msgs {
-		timestamp := timestamp(msg.Time)
-		crc32 := crc32OfMessage(magicByte, attributes, timestamp, msg.Key, msg.Value)
-		size := 4 + // crc
-			1 + // magic byte
-			1 + // attributes
-			8 + // timestamp
-			sizeofBytes(msg.Key) +
-			sizeofBytes(msg.Value)
-
-		writeInt64(w, msg.Offset)
-		writeInt32(w, int32(size))
-		writeInt32(w, int32(crc32))
-		writeInt8(w, magicByte)
-		writeInt8(w, attributes)
-		writeInt64(w, timestamp)
-		writeBytes(w, msg.Key)
-		writeBytes(w, msg.Value)
+		writeMessage(w, msg.Offset, attributes, msg.Time, msg.Key, msg.Value)
 	}
 
 	return w.Flush()
 }
+
+func compress(codec CompressionCodec, msgs ...Message) ([]Message, error) {
+	buf := &bytes.Buffer{}
```

Comment:
How about calling `buf.Grow` with an estimate of the final size here? It would save lots of copies and reallocations of the internal buffer.